src/store/mod.rs

```diff
@@ -0,0 +1,240 @@
+pub struct StoreOptions {
+    /// File path of the storage file.
+    file_path: PathBuf,
+    /// Max number of pages supported
+    pages_capacity: u64,
+    /// Maximum number of entries supported in the io_uring submission queue
+    io_uring_capacity: u32,
+    /// Reset the store to an empty state
+    reset: bool,
+}
+
+impl Default for StoreOptions {
+    fn default() -> Self {
+        Self {
+            file_path: PathBuf::from("node_pages_store"),
+            pages_capacity: 1_000,
+            io_uring_capacity: 512,
+            reset: true,
+        }
+    }
+}
+
+impl Store {
+    /// Create a new Store given the StoreOptions
+    pub fn new(options: StoreOptions) -> Self {
+        if options.reset {
+            // TODO: Is there a better alternative? fallocate, trucate or head maybe
+            let _ = std::process::Command::new("dd")
+                .args([
+                    "if=/dev/zero", // fill with zeros
+                    format!("of={}", options.file_path.to_string_lossy()).as_str(),
+                    format!("bs={}", options.pages_capacity).as_str(),
+                    format!("count={}", PAGE_SIZE).as_str(),
+                ])
+                .output()
+                .expect("Error creating empty store file");
+        }
+
+        let store_file = std::fs::File::options()
+            .write(true)
+            .read(true)
+            .open(options.file_path)
+            .expect("Imp open file");
+
+        let ring = Arc::new(
+            IoUring::<squeue::Entry, cqueue::Entry>::builder()
+                // TODO: Test those options
+                //.setup_sqpoll()
+                //.setup_iopoll()
+                .build(options.io_uring_capacity)
+                .expect("Error building io_uring"),
+        );
+
+        let (tx_new_batches, rx_new_batches) =
+            unbounded::<(Range<u64>, Option<Vec<Box<Page>>>, Sender<()>)>();
+        let (tx_kill_completion_handler, rx_kill_completion_handler) = bounded::<()>(1);
+
+        completion::start_completion_handler(
+            ring.clone(),
+            options.io_uring_capacity,
+            rx_new_batches,
+            rx_kill_completion_handler,
+        );
+
+        Store {
+            store_file,
+            capacity: options.pages_capacity,
+            ring_capacity: options.io_uring_capacity,
+            ring,
+            tx_new_batches,
+            tx_kill_completion_handler,
+            nonce: 0,
+        }
+    }
+
+    /// Write the pages at the specified index,
+    /// returns a WriteHandle with the number of written pages
+    /// or None if there was no space in the Submission queue to perform any write.
+    ///
+    /// Panics if the specified index is greater than the capacity of the storage.
+    pub fn write(&mut self, pages: &[(Page, u64)]) -> Option<(WriteHandle, u64)> {
+        let init_range = self.nonce;
+
+        let mut submissions = unsafe { self.ring.submission_shared() };
```

Comment:
Yes! But if I'm not mistaken, then you need to handle lifetimes, because that's a borrow

so I preferred to go the easier path, for sure a SAFETY note would be totally required